lib/auth/idemeum.go

```diff
@@ -0,0 +1,149 @@
+package auth
+
+import (
+	"context"
+	"github.com/gravitational/teleport/lib/idemeumjwt"
+	"time"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/constants"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+)
+
+func (a *Server) ValidateIdemeumServiceToken(ctx context.Context, ServiceToken string, TenantUrl string) (types.WebSession, error) {
+	//validate idemeum token
+	log.Info("Validating idemeum service token")
+	params, err := validateIdemeumToken(ServiceToken, TenantUrl)
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	//create user in teleport
+	log.Info("Creating idemeum user in the system")
+	user, err := a.createIdemeumUser(params)
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	//issue web session only for now
+	log.Infof("Issuing web session for user: %v", user.GetName())
+	session, err := a.createWebSession(ctx, types.NewWebSessionRequest{
+		User:       user.GetName(),
+		Roles:      user.GetRoles(),
+		Traits:     user.GetTraits(),
+		SessionTTL: params.sessionTTL,
+		LoginTime:  a.clock.Now().UTC(),
+	})
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.Infof("Issued web session for user: %v", user.GetName())
+	log.Info("session id:%v, expires: %v", session.GetName(), session.GetExpiryTime())
+	return session, nil
+}
+
+func validateIdemeumToken(ServiceToken string, TenantUrl string) (*createUserParams, error) {
+	if ServiceToken == "" || TenantUrl == "" {
+		return nil, trace.BadParameter("missing service token or tenant url")
+	}
+
+	claims, err := idemeumjwt.ValidateJwtToken(ServiceToken, TenantUrl)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	//For now add the restriction so that normal users and admin cant use their own tokens
+	userName := claims.Subject
+	if userName == "system" {
+		return nil, trace.BadParameter("no `userName` in claims for Idemeum account")
+	}
+
+	roles := claims.Roles
+	if len(claims.Roles) == 0 {
```

Comment:
Validate Following.

1. Issuer must be : `https://<tenant>.idemeum[lab].com`
2. Audience must be : `https://<tenant>.remote.idemeum[lab].com`